cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaPool.java

```diff
@@ -0,0 +1,296 @@
+        ModelUtils.validateComputeResources(pool.getSpec().getResources(), "KafkaNodePool.spec.resources");
+        StorageUtils.validatePersistentStorage(pool.getSpec().getStorage(), "KafkaNodePool.spec.storage");
+
+        KafkaPool result = new KafkaPool(reconciliation, kafka, pool, componentName(kafka, pool), ownerReference, idAssignment);
+
+        result.gcLoggingEnabled = isGcLoggingEnabled(kafka, pool);
+        result.jvmOptions = pool.getSpec().getJvmOptions() != null ? pool.getSpec().getJvmOptions() : kafka.getSpec().getKafka().getJvmOptions();
+        result.resources = pool.getSpec().getResources() != null ? pool.getSpec().getResources() : kafka.getSpec().getKafka().getResources();
+
+        if (useKRaft)   {
+            // KRaft is enabled. The roles are what is set in the KafkaNodePool CR or mixed node by default
+            result.kraftRoles = pool.getSpec().getRoles() != null && !pool.getSpec().getRoles().isEmpty() ? new HashSet<>(pool.getSpec().getRoles()) : Set.of(ProcessRoles.CONTROLLER, ProcessRoles.BROKER);
+        } else {
+            // KRaft is disabled. The role is always only a broker!
+            result.kraftRoles = Set.of(ProcessRoles.BROKER);
+        }
+
+        if (oldStorage != null) {
+            Storage newStorage = pool.getSpec().getStorage();
+
+            StorageDiff diff = new StorageDiff(reconciliation, oldStorage, newStorage, idAssignment.current(), idAssignment.desired());
+
+            if (!diff.isEmpty()) {
+                LOGGER.warnCr(reconciliation, "Only the following changes to Kafka storage are allowed: " +
+                        "changing the deleteClaim flag, " +
+                        "adding volumes to Jbod storage or removing volumes from Jbod storage, " +
+                        "changing overrides to nodes which do not exist yet " +
+                        "and increasing size of persistent claim volumes (depending on the volume type and used storage class).");
+                LOGGER.warnCr(reconciliation, "The desired Kafka storage configuration in the KafkaNodePool custom resource {}/{} contains changes which are not allowed. As a " +
+                        "result, all storage changes will be ignored. Use DEBUG level logging for more information " +
+                        "about the detected changes.", pool.getMetadata().getNamespace(), pool.getMetadata().getName());
+
+                Condition warning = StatusUtils.buildWarningCondition("KafkaStorage",
+                        "The desired Kafka storage configuration in the KafkaNodePool custom resource " + pool.getMetadata().getNamespace() + "/" + pool.getMetadata().getName() + " contains changes which are not allowed. As a " +
+                        "result, all storage changes will be ignored. Use DEBUG level logging for more information " +
+                        "about the detected changes.");
+                result.warningConditions.add(warning);
+
+                result.setStorage(oldStorage);
```

Comment:
We did use setters for everything in the past and then removed them because 99% of them did not do anything -> it just generated a lot of dummy code doing nothing. I you want to introduce them everywhere again, then I think that is way beyond the scope of this PR.

Alternatively, I can also remove the `setStorage` method here and simply do the validation before every time I set it directly => that would not be so hard.

Reply:
I don't think we should standardize on relying on the caller to call a validation prior to setting a value: It's too easy to forget. 

It would be better to use getters and setters everywhere, precisely because it means there's a single place to add/fix validation code. But I hear your complaint about having to maintain those methods when 95% of the time they're just boilerplate. We could use Lombok's [@Getter and @Setter](https://projectlombok.org/features/GetterSetter) on the model classes. We already use Lombok in the `api` module for `equals()` and hashCode()`, so this wouldn't be a huge step.

More radically, I don't think the models actually have to be mutable. They're not changed once they've been returned from `fromCrd()`. If that's correct then we could refactor them to be `record`s, validating their fields in the constructor. To avoid the `fromCrd()`  needing many many local variables (prior to instantiating the model record) we could use sundrio to generate a builder. I'm not at all certain that this is actually a good idea, but I'd be interested to know just how much it makes you wince :slightly_smiling_face: 

In any case, I'm not suggesting you need to do either of these changes in _this_ PR, merely that I don't think this is a pattern we actually want to use long term/promote.

Reply:
> It would be better to use getters and setters everywhere, precisely because it means there's a single place to add/fix validation code. But I hear your complaint about having to maintain those methods when 95% of the time they're just boilerplate. We could use Lombok's [@Getter and @Setter](https://projectlombok.org/features/GetterSetter) on the model classes. We already use Lombok in the api module for equals() and hashCode()`, so this wouldn't be a huge step.

I do not think the problem is in maintaining these methods. The problem is the code readability because you do not know what such methods actually do. When you have just an assignment, you know it is just an assignment and nothing more. If you have setters, you need to always check what are they actually doing - is there any additional logic? Or is there just an assignment? So the code becomes very unreadable as you need to jump into the implementations of the setters / getters to see what they are doing.

I also wonder about the memory consumption impact, since cleaning all these methods in the past was one of the things which seemed to help us void some of the OoM issues.

> More radically, I don't think the models actually have to be mutable. They're not changed once they've been returned from fromCrd(). If that's correct then we could refactor them to be records, validating their fields in the constructor. To avoid the fromCrd() needing many many local variables (prior to instantiating the model record) we could use sundrio to generate a builder. I'm not at all certain that this is actually a good idea, but I'd be interested to know just how much it makes you wince 🙂

I think this makes some sense. Most of the fields should be essentially final. They are not final only because we do not set them in the constructor but in the `fromCrd` method.

Reply:
Ok, I opened #8658 for this.